code/modules/reagents/reagent_containers/glass.dm

```diff
@@ -325,20 +296,109 @@
 	materials = list(/datum/material/glass=0)
 	volume = 50
 	amount_per_transfer_from_this = 10
+	fill_icon_thresholds = list(0, 10, 25, 50, 75, 80, 90)
+
+	// The 2 bottles have separate cap overlay icons because if the bottle falls over while bottle flipping the cap stays fucked on the moved overlay
+	var/cap_icon_state = "bottle_cap_small"
+	var/cap_on = TRUE
+	var/cap_lost = FALSE
+	var/mutable_appearance/cap_overlay
+	var/flip_chance = 10
+
+/obj/item/reagent_containers/glass/waterbottle/Initialize()
+	. = ..()
+	cap_overlay = mutable_appearance(icon, cap_icon_state)
+	if(cap_on)
+		spillable = FALSE
+		add_overlay(cap_overlay, TRUE)
+
+/obj/item/reagent_containers/glass/waterbottle/examine(mob/user)
+	. = ..()
+	if(cap_lost)
+		. += "<span class='notice'>The cap seems to be missing.</span>"
+	else if(cap_on)
+		. += "<span class='notice'>The cap is firmly on to prevent spilling. Alt-click to remove the cap.</span>"
+	else
+		. += "<span class='notice'>The cap has been taken off. Alt-click to put a cap on.</span>"
+
+/obj/item/reagent_containers/glass/waterbottle/AltClick(mob/user)
+	. = ..()
+	if(cap_lost)
+		to_chat(user, "<span class='warning'>The cap seems to be missing! Where did it go?</span>")
+		return
+	
+	var/fumbled = HAS_TRAIT(user, TRAIT_CLUMSY) && prob(5)
+	if(cap_on || fumbled)
+		cap_on = FALSE
+		spillable = TRUE
+		cut_overlay(cap_overlay, TRUE)
+		animate(src, transform = null, time = 2, loop = 0)
+		if(fumbled)
+			to_chat(user, "<span class='warning'>You fumble with [src]'s cap! The cap falls onto the ground and simply vanishes. Where the hell did it go?</span>")
+			cap_lost = TRUE
+		else
+			to_chat(user, "<span class='notice'>You remove the cap from [src].</span>")
+	else
+		cap_on = TRUE
+		spillable = FALSE
+		add_overlay(cap_overlay, TRUE)
+		to_chat(user, "<span class='notice'>You put the cap on [src].</span>")
+	update_icon()
+
+/obj/item/reagent_containers/glass/waterbottle/is_refillable()
+	if(cap_on)
+		return FALSE
+	. = ..()
+
+/obj/item/reagent_containers/glass/waterbottle/is_drainable()
+	if(cap_on)
+		return FALSE
+	. = ..()
+
+/obj/item/reagent_containers/glass/waterbottle/attack(mob/M, mob/user, obj/target)
+	if(cap_on && reagents.total_volume && istype(M))
+		to_chat(user, "<span class='warning'>You must remove the cap before you can do that!</span>")

```

Comment:
is this suppose to early return?

Reply:
I can make this early return but no code gets executed in `glass`' `attack()` or `afterattack()` because of a `spillable` check